### PR TITLE
Restormer Implementation

### PR DESCRIPTION
**Pull request by phisanti**
Fixes # .

### Description

This PR implements the Restormer architecture for high-resolution image restoration in MONAI following the discussion in issue #8261. The implementation supports both 2D and 3D images using MONAI's convolution as the base. Key additions include:

- Downsample class for efficient downsampling operations
- pixel_unshuffle operation complementing existing pixel_shuffle
- Channel Attention Block (CABlock) with FeedForward layer
- Multi-DConv Head Transposed Self-Attention (MDTA)
- OverlapPatchEmbed class
- Comprehensive unit tests for all new components

The implementation follows MONAI's coding patterns and includes performance validations against native PyTorch operations where applicable. 

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/blocks/cablock.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        x1, x2 = self.dwconv(x).chunk(2, dim=1)
+        return self.project_out(F.gelu(x1) * x2)
+
+
+class CABlock(nn.Module):
+    """Multi-DConv Head Transposed Self-Attention (MDTA): Differs from standard self-attention
+    by operating on feature channels instead of spatial dimensions. Incorporates depth-wise
+    convolutions for local mixing before attention, achieving linear complexity vs quadratic
+    in vanilla attention. Based on SW Zamir, et al., 2022 <https://arxiv.org/abs/2111.09881>"""
+
```
Comment (ericspod) [inline monai/networks/blocks/cablock.py:72]: We should have a full docstring here describing the arguments for the constructor, and in the previous class.

File: monai/networks/nets/restormer.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+            out_channels=embed_dim,
+            kernel_size=3,
+            strides=1,
+            padding=1,
+            bias=bias,
+            conv_only=True,
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.proj(x)
```
Comment (ericspod) [inline monai/networks/nets/restormer.py:70]:
```suggestion
class OverlapPatchEmbed(Convolution):
    """
    Initial feature extraction using overlapped convolutions. Unlike standard patch embeddings 
    that use non-overlapping patches, this approach maintains spatial continuity through 3x3 convolutions.
    """

    def __init__(self, spatial_dims: int, in_c: int = 3, embed_dim: int = 48, bias: bool = False):
        super().__init__(
            spatial_dims=spatial_dims,
            in_channels=in_c,
            out_channels=embed_dim,
            kernel_size=3,
            strides=1,
            padding=1,
            bias=bias,
            conv_only=True,
        )
```
Would it work to inherit directly from `Convolution`?

**Review by ericspod — COMMENTED** [review]
Looks good overall but I had a few inline comments, and we should have full docstrings everywhere appropriate. For any classes meant for general purpose use (ie. not just by Restormer) please ensure they have docstring descriptions for the arguments (at the very least for constructor args). Thanks!

File: monai/networks/blocks/downsample.py [hunk 3]
```diff
@@ -61,3 +64,234 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        x = self.conv_block(x)
+        if not all(d % self.scale_factor == 0 for d in x.shape[2:]):
+            raise ValueError(
+                f"All spatial dimensions {x.shape[2:]} must be evenly " f"divisible by scale_factor {self.scale_factor}"
+            )
+        x = pixelunshuffle(x, self.dimensions, self.scale_factor)
+        return x
+
+
+Downsample = DownSample
```
Comment (aylward) [inline monai/networks/blocks/downsample.py:296]: @ericspod - do we normally provide alternative capitalizations to functions?   "Downsample" is the generally accepted term (vs "Down Sample" which is less common).

I suggest using "Downsample" throughout, unless we offer alternative usage elsewhere that I haven't encountered.   IDE auto-complete can help folks get the right capitalization.

Looking at your enums, you use "Downsample" which confirms (in my mind) that we should be using "Downsample" throughout.

File: monai/networks/nets/restormer.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        x = x + self.ffn(self.norm2(x))
+        return x
+
+
+class OverlapPatchEmbed(nn.Module):
+    """Initial feature extraction using overlapped convolutions.
+    Unlike standard patch embeddings that use non-overlapping patches,
+    this approach maintains spatial continuity through 3x3 convolutions."""
+
+    def __init__(self, spatial_dims: int, in_c: int = 3, embed_dim: int = 48, bias: bool = False):
```
Comment (aylward) [inline monai/networks/nets/restormer.py:56]: Spell-out in_c to in_channels

File: monai/networks/nets/restormer.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        - Encoder: Progressive feature downsampling with increasing channels
+        - Latent: Deep feature processing at lowest resolution
+        - Decoder: Progressive upsampling with skip connections
+        - Refinement: Final feature enhancement
+    """
+
+    def __init__(
+        self,
+        spatial_dims=2,
+        inp_channels=3,
```
Comment (aylward) [inline monai/networks/nets/restormer.py:92]: in_channels

File: monai/networks/nets/restormer.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        out_channels=3,
+        dim=48,
+        num_blocks=[1, 1, 1, 1],
+        heads=[1, 1, 1, 1],
+        num_refinement_blocks=4,
+        ffn_expansion_factor=2.66,
+        bias=False,
+        LayerNorm_type="WithBias",
```
Comment (aylward) [inline monai/networks/nets/restormer.py:100]: Make enum or convert to Bool (e.g., layer_norm_use_bias).

File: tests/test_CABlock.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        output = block(x)
+        # Channel attention: sum([1..8]) * (qkv_conv=1) * (dwconv=1) * (attn_weights=1) * (proj=1) = 36 * 2 = 72
+        expected = torch.full_like(x, 72.0)
+
+        assert_allclose(output, expected, atol=1e-6)
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (aylward) [inline tests/test_CABlock.py:141]: Looks good .... as long as the runtime is reasonable.

**Review by aylward — CHANGES_REQUESTED** [review]
Thank you for this outstanding contribution!

File: monai/networks/blocks/downsample.py (hunk 3 above)
Comment (ericspod) [reply monai/networks/blocks/downsample.py:296]: We have had this mechanism for other classes so other capitalisation could be used in scripts, such as "Transformd" and "TransformD". Here I'd say we don't need it though.

File: monai/networks/utils.py [hunk 8]
```diff
@@ -411,6 +412,48 @@ def pixelshuffle(x: torch.Tensor, spatial_dims: int, scale_factor: int) -> torch
...
+    """
+    dim, factor = spatial_dims, scale_factor
+    input_size = list(x.size())
+    batch_size, channels = input_size[:2]
+    scale_factor_mult = factor**dim
+    new_channels = channels * scale_factor_mult
+
+    if any(d % factor != 0 for d in input_size[2:]):
+        raise ValueError(
+            f"All spatial dimensions must be divisible by factor {factor}. " f"Got spatial dimensions: {input_size[2:]}"
```
[the remark this answers is not in the document]
Comment (BoeckLab) [reply monai/networks/utils.py]: Done

File: monai/networks/blocks/downsample.py [hunk 9]
```diff
@@ -61,3 +64,234 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+                        in_channels=in_channels, out_channels=out_channels or in_channels, kernel_size=1, bias=bias
+                    ),
+                )
+            self.add_module(
+                "avgpool", Pool[Pool.AVG, spatial_dims](kernel_size=kernel_size_, stride=scale_factor_, padding=padding)
+            )
+            if post_conv:
+                self.add_module("postconv", post_conv)
+
+        elif down_mode == "pixelunshuffle":
```
[the remark this answers is not in the document]
Comment (BoeckLab) [reply monai/networks/blocks/downsample.py:206]: done, but I used DownsampleMode.PIXEL**UN**SHUFFLE as in downsampling the `restormer` uses `pixel_unshuffling` while `pixel_shuffling` is reserved for upsampling.

File: monai/networks/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (phisanti) [reply monai/networks/utils.py]: Done

File: monai/networks/blocks/downsample.py (hunk 3 above)
Comment (phisanti) [reply monai/networks/blocks/downsample.py:296]: Here, I mirrored the naming style of the Upsample class. Happy either way to remove it or keep it:

[monai/networks/blocks/upsample.py](https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/blocks/upsample.py#L300)

```python

Upsample = UpSample
Subpixelupsample = SubpixelUpSample = SubpixelUpsample


```

File: monai/networks/blocks/cablock.py (hunk 1 above)
Comment (phisanti) [reply monai/networks/blocks/cablock.py]: Done

File: monai/networks/nets/restormer.py (hunk 5 above)
Comment (phisanti) [reply monai/networks/nets/restormer.py]: Done

File: monai/networks/nets/restormer.py (hunk 4 above)
Comment (phisanti) [reply monai/networks/nets/restormer.py]: Done

File: monai/networks/blocks/downsample.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ericspod) [reply monai/networks/blocks/downsample.py]: I missed that, the value should be `PIXELUNSHUFFLE` indeed.

File: monai/networks/blocks/downsample.py (hunk 3 above)
Comment (ericspod) [reply monai/networks/blocks/downsample.py:300]: I'm fine with this included.

File: monai/networks/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ericspod) [reply monai/networks/utils.py]: I don't think this change took? It looks as before.

File: monai/networks/nets/restormer.py (hunk 4 above)
Comment (ericspod) [reply monai/networks/nets/restormer.py]: I think this hasn't taken as well.

**Comment by phisanti** [comment]
Hi @ericspod, no probs. Thanks for the feedback. I will move the test to the appropriate new folder this Friday. Also, I would like to ask for a bit of help if possible. It seems some of the tests cannot be running due to the optional import lineups. For example, it happens in `premerge-min/min-dep-py3 (3.12)` but not in `premerge/quick-py3 (ubuntu-latest)`. I tried to place the skip test on top, but maybe you can give me some indication of where it should go. 

Regarding the mixup on the changes, I think it happened when I did a merge back on Feb 7th as the branch got behind, I will try to look on it.

**Comment by ericspod** [comment]
Hi @phisanti We can look at the issues with the tests, no problem. I think I should let you figure out the mixup with branches and update your tests first then we can have a look at what the issue is. We should avoid just ignoring tests as much as possible, but if you need einops for example you can use `@skipUnless(has_einops, "Requires einops")` as a test method decorator. If that isn't working for you we can figure it out together.

**Comment by phisanti** [comment]
> Hi @phisanti We can look at the issues with the tests, no problem. I think I should let you figure out the mixup with branches and update your tests first then we can have a look at what the issue is. We should avoid just ignoring tests as much as possible, but if you need einops for example you can use `@skipUnless(has_einops, "Requires einops")` as a test method decorator. If that isn't working for you we can figure it out together.

Hi @ericspod, I finally manage to find some time to do the merge and be up-to-date with main. Now, in regards with the unit test, I placed them all in their appropriate folders mirroring the structure within source:
```
/tests/networks/utils/text_pixelshuffle.py
/tests/networks/blocks/text_CABlock.py
/tests/networks/blocks/text_downsample.py
/tests/networks/nets/text_restormer.py
```

I also noticed that I have to move the import of `SkipIfBeforePyTorchVersion` in `test_CABlock.py` from `tests.utils` to `tests.test_utils`. I have placed the statement `einops, has_einops = optional_import("einops")` plus     `@skipUnless(has_einops, "Requires einops")`, however, it seems some of the checks still fail. Thus, any help will be welcome, I believe the required changes are quite minimal.

**Comment by phisanti** [comment]
@ericspod  and @aylward, Thanks for your patience with this PR. It seems like the unit tests were creating a barrier to moving forward. The solution was actually quite simple - just needed to change the order of the SkipUnless decorator (since lower decorators execute first). You can now see that all the checks pass (except DCO due to some unsigned commits by other authors)

Does this address all the concerns? How would you like me to finalize these changes before the merge?

**Comment by ericspod** [comment]
Hi @phisanti We've done some updates to requirements and our testing process. I've updated your branch now so we'll see if things get broken, but if not what's left is to sort the DCO issue. The instruction you're given should be correct for doing a remedial commit, but if not we can work that out.

**Comment by phisanti** [comment]
Hi @ericspod, I'm glad to hear the code passes the checks! Regarding the DCO issue, I notice it requires specific remediation commits from Eric Kerfoot and Yiheng Wang. Since I'm not either of these authors, I don't think I can directly address this DCO requirement myself. What would be the appropriate next steps here to get this resolved?

**Comment by ericspod** [comment]
I will just mark the DCO as passed, this gets confused sometimes by commiting suggestions and other changes. @aylward your review is marked as "Request Changes", is there anything more you wanted to address? @KumoLiu Please trigger blossom when you can.
